File: java/FusedCANcoder/src/main/java/frc/robot/Robot.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot;

import com.ctre.phoenix6.BaseStatusSignal;
import com.ctre.phoenix6.StatusSignal;
import com.ctre.phoenix6.configs.CANcoderConfiguration;
import com.ctre.phoenix6.configs.TalonFXConfiguration;
import com.ctre.phoenix6.controls.DutyCycleOut;
import com.ctre.phoenix6.hardware.CANcoder;
import com.ctre.phoenix6.hardware.TalonFX;
import com.ctre.phoenix6.signals.AbsoluteSensorRangeValue;
import com.ctre.phoenix6.signals.FeedbackSensorSourceValue;
import com.ctre.phoenix6.signals.SensorDirectionValue;

import edu.wpi.first.wpilibj.TimedRobot;
import edu.wpi.first.wpilibj.XboxController;
import frc.robot.sim.PhysicsSim;

/**
 * The VM is configured to automatically run this class, and to call the functions corresponding to
 * each mode, as described in the TimedRobot documentation. If you change the name of this class or
 * the package after creating this project, you must also update the build.gradle file in the
 * project.
 */
public class Robot extends TimedRobot {
  private static final String canBusName = "canivore";
  private final TalonFX m_fx = new TalonFX(1, canBusName);
  private final CANcoder m_cc = new CANcoder(1, canBusName);

  private final StatusSignal<Boolean> f_fusedSensorOutOfSync = m_fx.getFault_FusedSensorOutOfSync();
  private final StatusSignal<Boolean> sf_fusedSensorOutOfSync = m_fx.getStickyFault_FusedSensorOutOfSync();
  private final StatusSignal<Boolean> f_remoteSensorInvalid = m_fx.getFault_RemoteSensorDataInvalid();
  private final StatusSignal<Boolean> sf_remoteSensorInvalid = m_fx.getStickyFault_RemoteSensorDataInvalid();

  private final StatusSignal<Double> fx_pos = m_fx.getPosition();
  private final StatusSignal<Double> fx_vel = m_fx.getVelocity();
  private final StatusSignal<Double> cc_pos = m_cc.getPosition();
  private final StatusSignal<Double> cc_vel = m_cc.getVelocity();
  private final StatusSignal<Double> fx_rotorPos = m_fx.getRotorPosition();

  private final DutyCycleOut m_dutyCycleControl = new DutyCycleOut(0);

  private final XboxController m_joystick = new XboxController(0);

  private int printCount = 0;

  private final Mechanisms m_mechanism = new Mechanisms();

  /**
   * This function is run when the robot is first started up and should be used for any
   * initialization code.
   */
  @Override
  public void robotInit() {
    /* Configure CANcoder to zero the magnet appropriately */
    CANcoderConfiguration cc_cfg = new CANcoderConfiguration();
    cc_cfg.MagnetSensor.AbsoluteSensorRange = AbsoluteSensorRangeValue.Signed_PlusMinusHalf;
    cc_cfg.MagnetSensor.SensorDirection = SensorDirectionValue.CounterClockwise_Positive;
    cc_cfg.MagnetSensor.MagnetOffset = 0.4;
    m_cc.getConfigurator().apply(cc_cfg);

    TalonFXConfiguration fx_cfg = new TalonFXConfiguration();
    fx_cfg.Feedback.FeedbackRemoteSensorID = m_cc.getDeviceID();
    fx_cfg.Feedback.FeedbackSensorSource = FeedbackSensorSourceValue.FusedCANcoder;
    fx_cfg.Feedback.SensorToMechanismRatio = 1.0;
    fx_cfg.Feedback.RotorToSensorRatio = 12.8;

    m_fx.getConfigurator().apply(fx_cfg);
  }

  @Override
  public void robotPeriodic() {
    if (printCount++ > 10) {
      printCount = 0;

      BaseStatusSignal.refreshAll(
        f_fusedSensorOutOfSync,
        sf_fusedSensorOutOfSync,
        f_remoteSensorInvalid,
        sf_remoteSensorInvalid,
        fx_pos, fx_vel,
        cc_pos, cc_vel);

      // If any faults happen, print them out. Sticky faults will always be present if live-fault occurs
      boolean anyFault = sf_fusedSensorOutOfSync.getValue() || sf_remoteSensorInvalid.getValue();
      if (anyFault) {
        System.out.println("A fault has occurred:");
        /* If we're live, indicate live, otherwise if we're sticky indicate sticky, otherwise do nothing */
        if (f_fusedSensorOutOfSync.getValue()) {
          System.out.println("Fused sensor out of sync live-faulted");
        } else if (sf_fusedSensorOutOfSync.getValue()) {
          System.out.println("Fused sensor out of sync sticky-faulted");
        }
        /* If we're live, indicate live, otherwise if we're sticky indicate sticky, otherwise do nothing */
        if (f_remoteSensorInvalid.getValue()) {
          System.out.println("Missing remote sensor live-faulted");
        } else if (sf_remoteSensorInvalid.getValue()) {
          System.out.println("Missing remote sensor sticky-faulted");
        }
      }

      if (m_joystick.getAButton()) {
        /* Clear sticky faults */
        m_fx.clearStickyFaults();
      }

      /* Print out current position and velocity */
      System.out.println("FX Position: " + fx_pos + " FX Vel: " + fx_vel);
      System.out.println("CC Position: " + cc_pos + " CC Vel: " + cc_vel);
      System.out.println("");
    }
    m_mechanism.update(fx_rotorPos, cc_pos, fx_pos);
  }

  @Override
  public void autonomousInit() {}

  @Override
  public void autonomousPeriodic() {}

  @Override
  public void teleopInit() {}

  @Override
  public void teleopPeriodic() {
    double output = m_joystick.getLeftY();
    if (Math.abs(output) < 0.1) output = 0;
    m_fx.setControl(m_dutyCycleControl.withOutput(output));
  }

  @Override
  public void disabledInit() {}

  @Override
  public void disabledPeriodic() {}

  @Override
  public void testInit() {}

  @Override
  public void testPeriodic() {}

  @Override
  public void simulationInit() {
    PhysicsSim.getInstance().addTalonFX(m_fx, m_cc, 12.8, 0.001);
  }

  @Override
  public void simulationPeriodic() {
    PhysicsSim.getInstance().run();
  }
}

File: cpp/FusedCANcoder/src/main/cpp/Robot.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "Robot.h"
#include <iostream>

using namespace ctre::phoenix6;

void Robot::RobotInit() {
  /* Configure CANcoder to zero the magnet appropriately */
  configs::CANcoderConfiguration cc_cfg{};
  cc_cfg.MagnetSensor.AbsoluteSensorRange = signals::AbsoluteSensorRangeValue::Signed_PlusMinusHalf;
  cc_cfg.MagnetSensor.SensorDirection = signals::SensorDirectionValue::CounterClockwise_Positive;
  cc_cfg.MagnetSensor.MagnetOffset = 0.4;
  m_cc.GetConfigurator().Apply(cc_cfg);

  configs::TalonFXConfiguration fx_cfg{};
  fx_cfg.Feedback.FeedbackRemoteSensorID = m_cc.GetDeviceID();
  fx_cfg.Feedback.FeedbackSensorSource = signals::FeedbackSensorSourceValue::FusedCANcoder;
  fx_cfg.Feedback.SensorToMechanismRatio = 1.0;
  fx_cfg.Feedback.RotorToSensorRatio = 12.8;

  m_fx.GetConfigurator().Apply(fx_cfg);
}
void Robot::RobotPeriodic() {
  if (printCount++ > 10) {
    printCount = 0;

    BaseStatusSignal::RefreshAll(
      f_fusedSensorOutOfSync,
      sf_fusedSensorOutOfSync,
      f_remoteSensorInvalid,
      sf_remoteSensorInvalid,
      fx_pos, fx_vel,
      cc_pos, cc_vel);

    // If any faults happen, print them out. Sticky faults will always be present if live-fault occurs
    bool anyFault = sf_fusedSensorOutOfSync.GetValue() || sf_remoteSensorInvalid.GetValue();
    if(anyFault) {
      std::cout << "A fault has occurred:" << std::endl;;
      /* If we're live, indicate live, otherwise if we're sticky indicate sticky, otherwise do nothing */
      if (f_fusedSensorOutOfSync.GetValue()) {
        std::cout << "Fused sensor out of sync live-faulted" << std::endl;
      } else if (sf_fusedSensorOutOfSync.GetValue()) {
        std::cout << "Fused sensor out of sync sticky-faulted" << std::endl;
      }
      /* If we're live, indicate live, otherwise if we're sticky indicate sticky, otherwise do nothing */
      if (f_remoteSensorInvalid.GetValue()) {
        std::cout << "Missing remote sensor live-faulted" << std::endl;
      } else if (sf_remoteSensorInvalid.GetValue()) {
        std::cout << "Missing remote sensor sticky-faulted" << std::endl;
      }
    }

    if (m_joystick.GetAButton()) {
      /* Clear sticky faults */
      m_fx.ClearStickyFaults();
    }

    /* Print out current position and velocity */
    std::cout << "FX Position: " << fx_pos << " FX Vel: " << fx_vel << std::endl;
    std::cout << "CC Position: " << cc_pos << " CC Vel: " << cc_vel << std::endl;
    std::cout << "" << std::endl;
  }
}

void Robot::AutonomousInit() {}
void Robot::AutonomousPeriodic() {}

void Robot::TeleopInit() {}
void Robot::TeleopPeriodic() {
  double output = m_joystick.GetLeftY();
  if (fabs(output) < 0.1) output = 0;
  m_fx.SetControl(m_dutyCycleControl.WithOutput(output));
}

void Robot::DisabledInit() {}
void Robot::DisabledPeriodic() {}

void Robot::TestInit() {}
void Robot::TestPeriodic() {}

void Robot::SimulationInit() {}
void Robot::SimulationPeriodic() {}

#ifndef RUNNING_FRC_TESTS
int main() {
  return frc::StartRobot<Robot>();
}
#endif
